trpcultivate_phenotypes/src/Plugin/Validators/validDataFile.php

```diff
@@ -0,0 +1,197 @@
+ *   validator_name = @Translation("Valid Data File Validator"),
+ *   input_types = {"file"}
+ * )
+ */
+class ValidDataFile extends TripalCultivatePhenotypesValidatorBase implements ContainerFactoryPluginInterface {
+  
+  /**
+   * This validator requires the following validator traits:
+   * - FileTypes: Gets an array of all supported MIME types the importer is configured to process.
+   */
+  use FileTypes;
+  
+  /**
+   * Entity Type Manager service.
+   * 
+   * @var EntityTypeManagerInterface
+   */
+  protected EntityTypeManagerInterface $service_EntityTypeManager;
+
+  /**
+   * Constructor.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, EntityTypeManagerInterface $entity_type_manager) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+  
+    // Set the Entity type manager service.
+    $this->service_EntityTypeManager = $entity_type_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition){
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('entity_type.manager')
+    );
+  }
+
+  /**
+   * Perform validation of data file.
+   * Checks include:
+   *  - Parameter filename or file id is valid.
+   *  - Has Drupal File Id number assigned and can be loaded.
+   *  - Valid file extension and file mime type configured by the file importer instance.
+   *  - File exists and not empty.
+   *  - File can be opened.
+   * 
+   * @param string $filename
+   *   The location of a file within the file system. 
+   * @param integer $fid
+   *   The unique identifier (fid) of a file that is managed by Drupal File System.    
+   * 
+   * @return array
+   *   An associative array with the following keys.
+   *    - case: a developer focused string describing the case checked.
+   *    - valid: either TRUE or FALSE depending on if the file is valid or not.
+   *    - failedItems: an array of "items" that failed to be used in the message to the user. This is an empty array if the file input was valid.
+   */
+  public function validateFile($filename, $fid = NULL) {
+    // Validator response values for a valid file.
+    $case = 'Data file is valid';
+    $valid = TRUE;
+    $failed_items = [];
+
+    // Parameter check, verify that the filename/file path is valid.
+    if (empty($filename) && is_null($fid)) {
+      return [
+        'case' => 'Filename is empty',
+        'valid' => FALSE,
+        'failedItems' => ['filename' => $filename]
+      ];
+    }
+    
+    // Parameter check, verify the file id number is not 0 or negative values.
+    if (!is_null($fid) && $fid <= 0) {
+      return [
+        'case' => 'Invalid file id number',
+        'valid' => FALSE,
+        'failedItems' => ['file_id' => $fid]
+      ];
+    } 
+
+    // File input.
+    $file_input = (is_null($fid)) ? $filename : $fid;
```

Comment:
Related to this, I agree we want to load the file based on fid and only based on filename if there is no fid.

However, I don't think this $file_input variable is helpful at all and I think they way we handle having both is to check that they match if the file entity is loaded via fid.

Instead I would do something like this to replace your lines 110-136:
```php
$file_entity = NULL;
if (is_numeric($fid) && $fid > 0) {
  // load the file entity based on the fid.
  // if not successful then failed case based on unloadable fid
  // else (i.e. loaded file entity) 
  //    check that the file entity->filename matches the filename passed into us.
  //    if they don't match then failure.
}
elseif ($filename) {
  // try to load file entity based on the filename.
  // if not successful then failed case based on filename not associated with file object.
}
// at this point $file_entity should be loaded or a failed case returned.
```

Note: you also don't need to know this key for saving to failedItems later. Instead you will have the file entity object and you should save both the filename and the fid to failedItems.

Reply:
My concerns here have been resolved ✅

Reply:
Unfortunately I still have concerns here... But Lacey and I discussed it a bit and think it's worth sitting down to re-think our use of parameters for this method. I created a new issue for this here: #110 so that it doesn't hold back merging in this PR